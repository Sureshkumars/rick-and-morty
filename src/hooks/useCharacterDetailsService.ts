import { useEffect, useState } from "react";
import { CharacterDetails } from "../types/characterDetail";
import { Service } from "../types/service";
import { CHARACTER_API_ENDPOINT } from "../constants/constants";
const characterApiEndPoint = CHARACTER_API_ENDPOINT;

const useCharacterDetailService = (characterId: string) => {
  //set the initial state
  const [result, setResult] = useState<Service<CharacterDetails>>({
    status: "loading",
  });
  useEffect(() => {
    const url = `${characterApiEndPoint}/${characterId}`;
    //API function to get the character details
    const fetchCharacterDetailsApi = async () => {
      try {
        // Fetch data from REST API
        const characterResponse = await (await fetch(url)).json();
        //Set the status to error in case of failure
        if (characterResponse.error) {
          setResult({
            status: "error",
            error: {
              name: "character-detail-error",
              message: "Unable to fetch character details",
            },
          });
        } else {
          //Reduce the episode array to single endpioint url with multiple episodes id in slug to get all the episdode details
          const epiSodeUrl: string = characterResponse.episode.reduce(
            (acc: string, url: string) => {
              return `${acc},${url.split(/[\s/]+/).pop()}`;
            }
          );
          //Handle the Episode API call based on the total episodes of the character
          const episodeResponse =
            characterResponse.episode.length !== 0
              ? await (await fetch(epiSodeUrl)).json()
              : [];
          if (episodeResponse.error) {
            setResult({
              status: "error",
              error: {
                name: "episode-detail-error",
                message: "Unable to fetch episode details",
              },
            });
          } else {
            //Handling for Single Episode response as it is returned as object
            const episodeDetails =
              characterResponse.episode.length === 1
                ? [
                    {
                      name: episodeResponse.name,
                      airedDate: episodeResponse.air_date,
                      season: episodeResponse.episode,
                    },
                  ]
                : episodeResponse.map((data: any) => {
                    return {
                      name: data.name,
                      airedDate: data.air_date,
                      season: data.episode,
                    };
                  });
            //Process the response data
            setResult({
              status: "loaded",
              payload: {
                name: characterResponse.name,
                imageUrl: characterResponse.image,
                lastLocation: characterResponse.location.name,
                totalEpisodes: characterResponse.episode.length,
                characterId: characterResponse.id,
                species: characterResponse.species,
                status: characterResponse.status,
                type: characterResponse.type,
                gender: characterResponse.gender,
                origin: characterResponse.origin.name,
                episode: episodeDetails,
                createdAt: characterResponse.created,
              },
            });
          }
        }
      } catch (error) {
        //Update the state to error
        setResult({ status: "error", error });
      }
    };

    // Call async function
    fetchCharacterDetailsApi();
  }, []);
  return result;
};

export default useCharacterDetailService;
