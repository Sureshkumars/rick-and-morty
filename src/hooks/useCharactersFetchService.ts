import { useEffect, useState } from "react";
import { CharacterSearchResults } from "../types/character";
import { Service } from "../types/service";
import { CHARACTER_API_ENDPOINT } from "../constants/constants";

const characterApiEndPoint = CHARACTER_API_ENDPOINT;

const useCharactersFetchService = (
  searchTerm: string,
  page?: number,
  showAll: boolean = false
) => {
  //set the initial state
  const [result, setResult] = useState<Service<CharacterSearchResults>>({
    status: "loading",
  });
  useEffect(() => {
    //When user visits for first time, fetch service won't get called as per our usecase

    if (!searchTerm && !showAll) {
      return;
    }

    //Toggles between searching for all characters or filter based on input search term
    let url =
      searchTerm === ""
        ? `${characterApiEndPoint}/?page=${page}`
        : `${characterApiEndPoint}/?name=${searchTerm}&page=${page}`;
    // Define asynchronous function
    const fetchCharactersApi = async () => {
      try {
        // Fetch data from REST API
        const response = await (await fetch(url)).json();
        //When no characters found API sends response with error key
        if (response.error) {
          setResult({
            status: "loaded",
            payload: {
              totalCount: 0,
              totalPages: 0,
              characters: [],
            },
          });
        } else {
          //Process the response data to according to our defined character types
          setResult({
            status: "loaded",
            payload: {
              totalCount: response.info.count,
              totalPages: response.info.pages,
              characters: response.results.map((data: any) => {
                return {
                  name: data.name,
                  imageUrl: data.image,
                  lastLocation: data.location.name,
                  totalEpisodes: data.episode.length,
                  characterId: data.id,
                  species: data.species,
                };
              }),
            },
          });
        }
      } catch (error) {
        //Catches error and update the return value
        setResult({ status: "error", error });
      }
    };

    // Call async function
    fetchCharactersApi();
  }, [searchTerm, page, showAll]);
  return result;
};

export default useCharactersFetchService;
