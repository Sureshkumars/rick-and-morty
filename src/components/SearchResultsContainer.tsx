import React, { useState, useEffect } from "react";
import { Flex, VStack, Box, Grid, Text, Button } from "@chakra-ui/core";
import { CharacterCard } from "./CharacterCard";
import useCharactersFetchService from "../hooks/useCharactersFetchService";
import useDebounce from "../hooks/useDebounce";
import Paginate from "./pagination/Pagination";
import { NoResults } from "./NoResults";
import { SkeletonLoader } from "./SkeletonLoader";
import { PageError } from "./PageError";
interface IProps {
  searchTerm: string;
  showAll: boolean;
  onResetSearch: () => void;
}
export const SearchResultsContainer: React.FC<IProps> = (props: IProps) => {
  const [page, setpageNumber] = useState(1);
  //Custom character fetch service hook with debounce limit of 500ms
  const searchResults = useCharactersFetchService(
    useDebounce(props.searchTerm),
    Math.ceil(page / 4),
    props.showAll
  );
  //Reset Page Number when search term changes
  useEffect(() => {
    setpageNumber(1);
  }, [props.searchTerm]);

  return (
    <Box>
      {/* Loader State*/}
      {searchResults.status === "loading" && props.showAll && (
        <SkeletonLoader />
      )}
      {/* Returns No results when search results are empty */}
      {searchResults.status === "loaded" &&
        searchResults.payload.totalCount === 0 && <NoResults />}
      {/* Loads the Search Results */}
      {searchResults.status === "loaded" &&
        searchResults.payload.totalCount > 0 && (
          <VStack spacing={8}>
            <Grid>
              <Text fontSize="lg">
                Found {searchResults.payload.totalCount} characters,
              </Text>
              <Button
                colorScheme="teal"
                variant="ghost"
                onClick={props.onResetSearch}
              >
                Reset Search
              </Button>
            </Grid>
            <Flex flexDirection="row" flexWrap="wrap" justifyContent="center">
              {/* As we wanted to show five results only per page and API service doesn't have option for fetching based on count. So slicing the data based on the total default value which is 20 per page from the API */}
              {searchResults.payload.characters
                .slice(((page - 1) * 5) % 20, (((page - 1) * 5) % 20) + 5)
                .map((character: any) => (
                  <CharacterCard
                    character={character}
                    key={character.characerId}
                  />
                ))}
            </Flex>
            {/* Math.ceil calculation is for transforming the total apges returned from API service results to our state data */}
            <Paginate
              currentPage={page}
              pageCount={Math.ceil(searchResults.payload.totalCount / 5)}
              onPageClick={(page) => {
                setpageNumber(page);
                window.scrollTo(0, 0);
              }}
            />
          </VStack>
        )}
      {searchResults.status === "error" && <PageError />}
    </Box>
  );
};
