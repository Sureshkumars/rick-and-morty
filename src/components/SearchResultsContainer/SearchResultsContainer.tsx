import React, { useState, useEffect } from "react";
import { Flex, VStack, Box, Grid, Text, Button } from "@chakra-ui/core";
import { CharacterCard } from "../CharacterCard/CharacterCard";
import { SkeletonLoader } from "../SkeletonLoader/SkeletonLoader";
import { ErrorMessage } from "../ErrorMessage/ErrorMessage";
import { NO_RESULTS_TEXT, ERROR_MESSAGE } from "../../constants/constants";
import { Link } from "react-router-dom";
import useCharactersFetchService from "../../hooks/useCharactersFetchService";
import useDebounce from "../../hooks/useDebounce";
import Paginate from "../Paginate/Paginate";

interface IProps {
  searchTerm: string;
  showAll: boolean;
  onResetSearch: () => void;
}
export const SearchResultsContainer: React.FC<IProps> = (props: IProps) => {
  const [page, setpageNumber] = useState(1);

  //Custom character fetch service hook with debounce limit of 500ms
  //Math.Ceil Calculation will change its value once in every four pages as per our usecase for fetiching API in our custom hook.
  const searchResults = useCharactersFetchService(
    useDebounce(props.searchTerm),
    Math.ceil(page / 4),
    props.showAll
  );

  //Reset Page Number when search term changes
  useEffect(() => {
    setpageNumber(1);
  }, [props.searchTerm]);

  return (
    <Box>
      {/* Loader State*/}
      {searchResults.status === "loading" && props.showAll && (
        <SkeletonLoader />
      )}
      {/* Returns No results when search results are empty */}
      {searchResults.status === "loaded" &&
        searchResults.payload.totalCount === 0 && (
          <ErrorMessage message={NO_RESULTS_TEXT} showHomeCTA={false} />
        )}
      {/* Loads the Search Results */}
      {searchResults.status === "loaded" &&
        searchResults.payload.totalCount > 0 && (
          <VStack spacing={8}>
            <Grid>
              <Text fontSize="lg">
                Found {searchResults.payload.totalCount} characters,
              </Text>
              <Button
                colorScheme="teal"
                variant="ghost"
                onClick={props.onResetSearch}
              >
                Reset Search
              </Button>
            </Grid>
            <Flex flexDirection="row" flexWrap="wrap" justifyContent="center">
              {/* As we wanted to show five results only per page and API service doesn't have option for fetching based on count. So slicing the data based on the total default value which is 20 per page from the API */}
              {searchResults.payload.characters
                .slice(((page - 1) * 5) % 20, (((page - 1) * 5) % 20) + 5)
                .map((character: any, i) => (
                  <Link
                    to={`/character/${character.characterId}`}
                    title={character.name}
                    key={i}
                  >
                    <CharacterCard character={character} />
                  </Link>
                ))}
            </Flex>
            {/* Math.ceil calculation is for transforming the total pages returned from API service results to our state data */}
            <Paginate
              currentPage={page}
              pageCount={Math.ceil(searchResults.payload.totalCount / 5)}
              onPageClick={(page) => {
                setpageNumber(page);
                window.scrollTo(0, 0);
              }}
            />
          </VStack>
        )}
      {searchResults.status === "error" && (
        <ErrorMessage message={ERROR_MESSAGE} showHomeCTA={true} />
      )}
    </Box>
  );
};
