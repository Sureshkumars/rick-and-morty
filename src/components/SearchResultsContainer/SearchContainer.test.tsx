import React from "react";
import { SearchResultsContainer } from "./SearchResultsContainer";
import { render } from "@testing-library/react";
import * as hooks from "../../hooks/useCharactersFetchService";
import { MemoryRouter, Route } from "react-router-dom";

describe("SearchResultsContainer", () => {
  Object.defineProperty(window, "matchMedia", {
    writable: true,
    value: jest.fn().mockImplementation((query) => ({
      matches: false,
      media: query,
      onchange: null,
      addListener: jest.fn(), // deprecated
      removeListener: jest.fn(), // deprecated
      addEventListener: jest.fn(),
      removeEventListener: jest.fn(),
      dispatchEvent: jest.fn(),
    })),
  });

  it("matches snapshot", () => {
    jest.spyOn(hooks, "default").mockImplementation(() => ({
      status: "loaded",
      payload: {
        totalCount: 2,
        totalPages: 1,
        characters: [
          {
            name: "Rick",
            imageUrl: "/images/character-fallback.jpeg",
            species: "Alien",
            lastLocation: "Earth",
            totalEpisodes: 22,
            characterId: 1,
          },
          {
            name: "Morty",
            imageUrl: "/images/character-fallback.jpeg",
            species: "Human",
            lastLocation: "Earth Dimension 1",
            totalEpisodes: 10,
            characterId: 3,
          },
        ],
      },
    }));

    const { asFragment }: any = render(
      <MemoryRouter>
        <Route>
          <SearchResultsContainer
            searchTerm="Rick"
            showAll={true}
            onResetSearch={() => {
              //Search Reset will be triggered.
            }}
          />
        </Route>
      </MemoryRouter>
    );

    expect(asFragment()).toMatchSnapshot();
  });
});
