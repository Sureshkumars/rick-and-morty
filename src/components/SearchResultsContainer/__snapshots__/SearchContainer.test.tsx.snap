// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`SearchResultsContainer matches snapshot 1`] = `
<DocumentFragment>
  <div
    class="css-0"
  >
    <div
      class="chakra-stack css-pk93zj"
    >
      <div
        class="css-lgj0h8"
      >
        <p
          class="chakra-text css-1ezsviu"
        >
          Found 2 characters,
        </p>
        <button
          class="chakra-button css-oqfhsh"
          type="button"
        >
          Reset Search
        </button>
      </div>
      <div
        class="css-1xg6dlq"
      >
        <a
          href="/character/1"
          title="Rick"
        >
          <div
            class="css-ao3w3m"
          >
            <img
              alt="Rick"
              class="chakra-image__placeholder css-0"
              src="/images/character-fallback.jpeg"
            />
            <div
              class="css-n16x53"
            >
              <div
                class="css-x1sij0"
              >
                <span
                  class="chakra-badge css-637omx"
                >
                  Alien
                </span>
              </div>
              <h4
                class="css-mvaft7"
              >
                <strong>
                  Rick 
                </strong>
              </h4>
              <p
                class="css-0"
              >
                <strong>
                  Last location - 
                </strong>
                 Earth
              </p>
              <div
                class="css-15yujhg"
              >
                <span
                  class="css-bgvdb6"
                >
                  Totally 22 episodes
                </span>
              </div>
            </div>
          </div>
        </a>
        <a
          href="/character/3"
          title="Morty"
        >
          <div
            class="css-ao3w3m"
          >
            <img
              alt="Morty"
              class="chakra-image__placeholder css-0"
              src="/images/character-fallback.jpeg"
            />
            <div
              class="css-n16x53"
            >
              <div
                class="css-x1sij0"
              >
                <span
                  class="chakra-badge css-637omx"
                >
                  Human
                </span>
              </div>
              <h4
                class="css-mvaft7"
              >
                <strong>
                  Morty 
                </strong>
              </h4>
              <p
                class="css-0"
              >
                <strong>
                  Last location - 
                </strong>
                 Earth Dimension 1
              </p>
              <div
                class="css-15yujhg"
              >
                <span
                  class="css-bgvdb6"
                >
                  Totally 10 episodes
                </span>
              </div>
            </div>
          </div>
        </a>
      </div>
      <div
        class="paginationContainer"
      >
        <div
          class="paginationButtonContainer"
        >
          <div
            class="pagination_button highlight "
          >
            1
          </div>
        </div>
      </div>
    </div>
  </div>
</DocumentFragment>
`;
